### PR TITLE
Sparse poly implement poly

### PR DESCRIPTION
**Pull request by ValarDragon**
## Description

This PR makes SparsePolynomial implement Polynomial, and adds several more tests to SparsePolynomial.

It also adds a paralell implementation of `SparsePolynomial.evaluate()`.

closes: #116 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by Pratyush** [comment]
Hmm is there a place where the parallel evaluation of sparse poly would help? The use case I had in mind was polynomials similar to the vanishing polynomial for multiplicative subgroups, where parallel evaluation would definitely hurt

**Comment by ValarDragon** [comment]
Ah thats fair. I was thinking that we should probably special case regardless the case where all indices are powers of two, to not repeat squaring terms. (This captures vanishing polynomials of the union of cosets, and the vp for binary fields)

I was at the time of writing thinking that this would be more important here at lower numbers of terms than it would for the case of dense polynomial evaluation, since each term here recomputes the entire power. 

Though I think that really that tells us that we should just compute all the squares first, and then just have additions remaining for each term. (And then as shown in #119, not really a need to parallelize the additions)

I'll remove parallelization from evaluate for now, in favor of eventually doing the above types of methods.

File: poly/src/domain/radix2.rs [hunk 1]
```diff
@@ -225,11 +228,13 @@ pub(crate) fn serial_radix2_fft<T: DomainCoeff<F>, F: FftField>(a: &mut [T], ome
     }
 
     let mut m = 1;
-    for _ in 0..log_n {
+    for _i in 1..=log_n {
```
Comment (Pratyush) [inline poly/src/domain/radix2.rs:231]: Hmm so the old code was subtly wrong?

Comment (Pratyush) [reply poly/src/domain/radix2.rs:231]: Or no, because we didn't use the index anywhere.

Comment (ValarDragon) [reply poly/src/domain/radix2.rs:231]: It wasn't wrong, but writing the invariant to understand the code was awkward with the prior bounds

**Review by Pratyush — APPROVED** [review]

**Pull request merged**